*** MemoryHeapBase.h	Wed Nov 28 16:00:32 2018
--- MemoryHeapBase.hh	Wed Nov 28 16:00:57 2018
*************** protected:
*** 86,92 ****
              int flags = 0, const char* device = NULL);
  
  private:
+     status_t mapion(int fd, size_t size, uint32_t offset = 0);
      status_t mapfd(int fd, size_t size, uint32_t offset = 0);
  
      int         mFD;
--- 86,91 ----
*************** private:
*** 96,102 ****
      const char* mDevice;
      bool        mNeedUnmap;
      uint32_t    mOffset;
+     int         mFD2;
  };
  
  // ---------------------------------------------------------------------------
--- 95,100 ----
